### PR TITLE
#94: Extract logic for TgzArchive metadata

### PR DESCRIPTION
**Pull request by andreoss**
For #94 
- Extract method
- Refactor dependent classes

### CODE REVIEW

**Review by genryxy — CHANGES_REQUESTED** [review]
@andreoss thanks, see comments

File: src/main/java/com/artipie/helm/metadata/IndexYamlMapping.java [hunk 1]
```diff
@@ -114,37 +116,32 @@ public IndexYamlMapping(final Map<String, Object> mapfromindex) {
...
+                            chart -> chart.get(IndexYamlMapping.VRSN)
+                                .equals(opt.get().get(IndexYamlMapping.VRSN))
+                        );
+                    }
+                    existed.add(vers);
                 }
-                existed.add(vers);
+            } else {
+                entr.put(name, versions);
             }
```
Comment (genryxy) [inline src/main/java/com/artipie/helm/metadata/IndexYamlMapping.java:141]: @andreoss also, now synchronized section includes more logic for processing. Previously this section includes only check does mapping contain specific key. Is it possible to reduce the amount of logic in synchronized block?

File: src/main/java/com/artipie/helm/metadata/IndexYaml.java [hunk 2]
```diff
@@ -165,28 +161,28 @@ public Completable deleteByNameAndVersion(final String name, final String versio
...
+        final Map<String, Object> index,
+        final TgzArchive tgz
+    ) {
+        final Map<String, Object> copy = new HashMap<>(index);
+        final IndexYamlMapping yaml = new IndexYamlMapping(copy);
+        if (
+            !yaml
+                .byChartAndVersion(
+                    tgz.chartYaml().name(),
+                    tgz.chartYaml().version()
```
[the remark this answers is not in the document]
Comment (andreoss) [reply src/main/java/com/artipie/helm/metadata/IndexYaml.java:176]: @genryxy The existence check is in order to not update metadata twice.

File: src/test/java/com/artipie/helm/TgzArchiveTest.java [hunk 3]
```diff
@@ -91,4 +94,23 @@ public void savedCorrectly(@TempDir final Path tmp) throws IOException {
...
+            Matchers.allOf(
+                Matchers.hasEntry(
+                    Matchers.is("urls"),
+                    Matchers.is(Collections.singletonList("tomcat-0.4.1.tgz"))
+                ),
+                Matchers.hasEntry(
+                    Matchers.is("digest"),
+                    Matchers.isA(String.class)
+                )
+            )
```
[the remark this answers is not in the document]
Comment (andreoss) [reply src/test/java/com/artipie/helm/TgzArchiveTest.java]: @genryxy Done

File: src/main/java/com/artipie/helm/TgzArchive.java [hunk 4]
```diff
@@ -94,6 +97,30 @@ public String name() {
         return String.format("%s-%s.tgz", chart.name(), chart.version());
     }
 
+    /**
+     * Obtain metadata for Index.
+     * @param baseurl Base url.
+     * @return Metadata of archive.
+     */
```
[the remark this answers is not in the document]
Comment (andreoss) [reply src/main/java/com/artipie/helm/TgzArchive.java:104]: @genryxy

File: src/main/java/com/artipie/helm/TgzArchive.java [hunk 5]
```diff
@@ -94,6 +97,30 @@ public String name() {
...
+                Collections.singletonList(
+                    String.format(
+                        "%s%s",
+                        baseurl.orElse(""),
+                        this.name()
+                    )
+                )
+            )
+        );
+        meta.put("digest", this.digest());
```
[the remark this answers is not in the document]
Comment (andreoss) [reply src/main/java/com/artipie/helm/TgzArchive.java:119]: @genryxy Removed the method

File: src/main/java/com/artipie/helm/metadata/IndexYamlMapping.java (hunk 1 above)
Comment (andreoss) [reply src/main/java/com/artipie/helm/metadata/IndexYamlMapping.java:141]: @genryxy I left a todo to fix this.

File: src/test/java/com/artipie/helm/TgzArchiveTest.java [hunk 6]
```diff
@@ -91,4 +94,23 @@ public void savedCorrectly(@TempDir final Path tmp) throws IOException {
         );
         vertx.close();
     }
+
+    @Test
+    void hasCorrectMetadata() throws IOException {
+        MatcherAssert.assertThat(
+            new TgzArchive(
+                Files.readAllBytes(Paths.get("./src/test/resources/tomcat-0.4.1.tgz"))
+            ).metadata(Optional.empty()),
```
[the remark this answers is not in the document]
Comment (andreoss) [reply src/test/java/com/artipie/helm/TgzArchiveTest.java]: @genryxy Done

File: src/test/java/com/artipie/helm/TgzArchiveTest.java [hunk 7]
```diff
@@ -74,21 +82,40 @@ public void savedCorrectly(@TempDir final Path tmp) throws IOException {
...
-                )
+                file.asBytes()
             )
         );
         vertx.close();
     }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    void hasCorrectMetadata() {
```
Comment (genryxy) [inline src/test/java/com/artipie/helm/TgzArchiveTest.java:102]: @andreoss why do we need suppress warning here?

File: src/main/java/com/artipie/helm/metadata/IndexYaml.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (genryxy) [reply src/main/java/com/artipie/helm/metadata/IndexYaml.java]: @andreoss ok, got it

File: src/test/java/com/artipie/helm/TgzArchiveTest.java (hunk 7 above)
Comment (andreoss) [reply src/test/java/com/artipie/helm/TgzArchiveTest.java:102]: @genryxy There's an issue with type inference on JDK 8. It's not needed on JDK 11

**Review by genryxy — DISMISSED** [review]
@andreoss thanks, looks good to me
@g4s8 please check/merge

File: pom.xml [hunk 8]
```diff
@@ -165,6 +165,12 @@ SOFTWARE.
       <id>qulice</id>
       <build>
         <plugins>
+          <!--
+           @todo #94:30min Enable forbidden-apis plugin.
+            Check code for compliance with code style described
+            in CONTRIBUTING.md. Fix as many classes as time allows
+            and than update this todo to fix the rest.
+          -->
```
Comment (g4s8) [inline pom.xml:173]: @andreoss why it's related to #94 ticket? It doesn't seem to be a part of the task, so it should be reported as a new ticket, not as a todo for #94

**Review by g4s8 — CHANGES_REQUESTED** [review]
@andreoss please check my comment

File: pom.xml (hunk 8 above)
Comment (andreoss) [reply pom.xml:173]: @g4s8 It constantly comes up on review. Is there a difference between the issues reported via pdd or manually besides the issues will be linked?

Comment (g4s8) [reply pom.xml:173]: @andreoss yes, it allows to control the scope better by filtering different kind of tasks and building the tree of subtasks as dependencies, so we'll have better granularity of scope in that case.

**Comment by genryxy** [comment]
@andreoss could you please remove todo and it will be fine

**Review by genryxy — APPROVED** [review]
@andreoss thanks, approve
@g4s8 could you please check/merge

**Review by g4s8 — APPROVED** [review]
@andreoss thanks

**Comment by 0crat** [comment]
Job `gh:artipie/helm-adapter#101` is not assigned, can't get performer

**Pull request merged**